## print() function:-

    It is a function Prints the given objects to the standart output device. from 3.x version onwards
    
    sybtax:- 
    
        print( value/Expression , sep="", end="\n", file=sys.stdout, flush=False)
        
        

In [112]:
print(1,2,3)

1 2 3


In [68]:
print("1,2,3,4")

1,2,3,4


In [113]:
print(1234456)

1234456


In [77]:
print(2+3-10/5*3)

-1.0


In [78]:
print(20-10+40/10)

14.0


### sep=""

In [74]:
print("1234455", sep='@')

1234455


In [116]:
print(1,2,3,4,5, sep=" ")

1 2 3 4 5


In [118]:
print([1,2,3,5,84,2],[4,3,'Bangalore'], sep='!')

[1, 2, 3, 5, 84, 2]![4, 3, 'Bangalore']


In [75]:
print('python','machine learning', 'deep learning', sep=" @@@@@ ")

python @@@@@ machine learning @@@@@ deep learning


In [76]:
print('p','y','t','h','o','n', sep='______')

p______y______t______h______o______n


## end=''

In [81]:
print("python is very Easy Language")
print("Machine Language")

python is very Easy Language
Machine Language


In [122]:
print("python is very Easy Language",end=' ')
print("Machine Language")

python is very Easy Language,Machine Language


In [83]:
print("python is very Easy Language",end='\n\n\n')
print("Machine Language")

python is very Easy Language


Machine Language


In [85]:
print("python is very Easy Language\n",end=' ')
print("Machine Language")

python is very Easy Language
 Machine Language


                                 other programming                      in python
   
----------------------------------------------------------------------------------

       ' '      ----->           character                             string 
       " "      ------>          string                                string
       ''' '''  ------>          character string documentation        one/many lines
       """ """ -------->         string documentation                  one/many lines

In [130]:
print('Python is very easy language')

Python is very easy language


In [95]:
print('Python is very easy \
language')

Python is very easy language


In [96]:
print("Python is very easy language")

Python is very easy language


In [97]:
print("Python is very \
easy language")

Python is very easy language


In [92]:
print('''Python is very easy language''')

Python is very easy language


In [131]:
print('''Python is very easy 
language 
and machine learning''')

Python is very easy 
language
and machine learning


In [94]:
print(""" Machine learning is very demanding now """)

 Machine learning is very demanding now 


In [98]:
print(""" Machine learning is very
demanding now """)

 Machine learning is very
demanding now 


### space sequence ( back slash "\\")

In [134]:
print(" hi i am \"Jay Shankar Sharma\"")

 hi i am "Jay Shankar Sharma"


In [136]:
print(" hi i am 'Jay shankar sharma'")

 hi i am 'Jay shankar sharma'


In [102]:
print('"python" is very easy Language')

"python" is very easy Language


In [104]:
print('\'Bangarore\' is capital of karnataka')

'Bangarore' is capital of karnataka


In [105]:
print(""" 'Bangalore' is captial of karnataka""")

 'Bangalore' is captial of karnataka


In [107]:
print(''' "Bangalore" is captial of karnataka''')

 "Bangalore" is captial of karnataka


In [140]:
print("E:\\zipir_sniffer_20200810\\all_ids_old")

E:\zipir_sniffer_20200810\all_ids_old


In [141]:
print(r'E:\zipir_sniffer_20200810\all_ids_old')

E:\zipir_sniffer_20200810\all_ids_old


### Reading data From KeyBoard or User:-

     Python Provides Two Built-in function to read a line of text from standart input.
     
     syntax:-
     
         1. raw_input()   --------> 2.x version
         2. input()       --------> 3.x version

In [153]:
a = input()

123


In [155]:
type(int(a))

int

#### format() function:-
It returns a formatted representated of the given value controlled by the format specifier

syntax:-
    
    format(value[, firmat_spec[,)
    
    '<'  ------>  the field will be left-aligned within the avalible space
    '>' -------->  the field will be right-aligned within the avaliable space 
    '^' --------> the field will be centered within the avaliable space

In [196]:
print(format("1234", "s"))

1234


In [197]:
print(format(12345,'2d'))

12345


In [50]:
print(format(12345,'f'))

12345.000000


In [156]:
print(format(12345, '.2f'))

12345.00


In [52]:
print(format(3/4))

0.75


In [53]:
print(format(3/4,'f'))

0.750000


In [54]:
print(format(3/4,'.1f'))

0.8


In [158]:
# left alignment

print(format(1234,"<20"), end='')
print('a')

1234                a


In [56]:
# right laignment

print(format(1234,'>20'))

                1234


In [57]:
# center alignment

print(format(1234,"^20"))

        1234        


In [7]:
a = 10
b = 20

print(b," is greater than ",a )

print("{} is greater than {}".format(b,a))

print("{0} is greater than {1}".format(b,a))
print(a)

20  is greater than  10
20 is greater than 10
20 is greater than 1010


In [213]:
a = 106
b = 205

print(b," is greater than ", a)

print("{} is greater than {}".format(b,a))

print("{1} is greater than {0}".format(a,b))

205  is greater than  106
205 is greater than 106
205 is greater than 106


#### Single lone Underscore (_)
    the most recent output value is automatically stored by the interpreter in a special variable with the name "_".

In [177]:
a = 60 
b = 10
a+b

70

In [178]:
print(_)

70


In [179]:
d = 50

In [180]:
_-_

0

In [181]:
_

0

##### Variables:-
 A variable is a name that refers to a value , a variable points to a memory location where data is read and modify.
 
##### Identifier:-
  An Identifier is just a name of the variable.
  
  
##### Python variables names rules:-
   - Python Variable must begin with (a-z), (A-Z), _
   - python varaible must not be reserved word or keyword
   - python variavle name should be written in camelCase
   - Python variable name should be any resonable length

In [214]:
a = 10
a

10

#### constants:- 
    Fixed values such as numbers, letters , strings are Constant

#### literals:-
    A literal is a notation for representing a fixed values.
    
    Python supports following literals:-
    
        - string literals
        - Numeric literals --> int, long , float, complex
        - Boolean literals --> True , False
        - Special literals --> None
        - literals collection --> list , tuple, set, dictionary

In [39]:
# String Literals 

pystr = "PYTHON"

print("String Literals is :-", pystr)

String Literals is :- PYTHON


In [227]:
# decimal literals

d = bin(10)
print(d)

0b1010


In [232]:
b = 0B1011
print(b)

11


In [233]:
# Octal literals
octal = oct(23)
print(octal)

0o27


In [11]:
octal = oct(0o12)   #-----> 0 , 1, 2, 3, 4, 5, 6, 7 
print(octal)

0o12


In [234]:
# Hexadecimal Literals
hexadecimal = 1234
print(hex(hexadecimal))

0x4d2


In [192]:
# Integer literals 

Pyint = 1234 
print("integer Literals is :-", Pyint)

integer Literals is :- 1234


In [193]:
# Float literals

pyfloat = 1234.14784
print("Float literals is :- ", pyfloat)

Float literals is :-  1234.14784


In [198]:
# complex literals:-

PyCom = 2 + 3j

print("Complex literals :- ", PyCom)

Comples literals :-  (2+3j)


In [200]:
# Boolean Literals

PyBool = True

print("Boolean Literals :-", PyBool)

PyBool1 = False

print("Boolean Literals :-", PyBool1)

Boolean Literals :- True
Boolean Literals :- False


In [202]:
# specal Literals :- NoneType

PySec = None
print(type(PySec))
print("special liteals :-", PySec)

<class 'NoneType'>
special liteals :- None


In [ ]:
# literals Collection
# list, tuple, set, frozenset, dict

In [217]:
# List Collection 

pylist = [1,2,3]
print("List literals is ", pylist)

List literals is  [1, 2, 3]


In [219]:
# tuple Collection 

pytuple = (5,3,4)
print("tuple literals is ", pytuple)

tuple literals is  (5, 3, 4)


In [220]:
# set Collection 

pyset = {2,4,5,8,2}
print("set literals is ", pytuple)

set literals is  (5, 3, 4)


In [221]:
# frosenset Collection 

pyfrozenset = frozenset({2,4,5,8,2})
print("frozenset literals is ", pyfrozenset)

frozenset literals is  frozenset({8, 2, 4, 5})


In [222]:
# dictionary Collection 

pydict = {1:"one", 2:"Two"}
print("set literals is ", pydict)

set literals is  {1: 'one', 2: 'Two'}


#### DataTypes:-
   - integer
   - float
   - complex
   - string
   - list
   - tuple
   - set
   - Frozenset
   - Dictionary
   

How check the datatypes of object/variable
   - type() function:- 
        it returns the type of the given object

        syntax:
            type(obj)

In [30]:
# Integer Datatype

i = 10
print(i)
print(type(a))

10
<class 'int'>


In [29]:
# float data types

f = 12.90
print(f)
print(type(f))

12.9
<class 'float'>


In [28]:
# complex datatypes

c = 6 + 7j
print(c)
print(type(c))

(6+7j)
<class 'complex'>


In [17]:
# Long :- Deleted form python 3.0 onwards

In [27]:
# string 

s = "Python"
print(s)
print(type(s))

Python
<class 'str'>


In [10]:
# Empty list

l = []
print(type(l))

<class 'list'>


In [26]:
l = [1,5,6,9]
print(l)
print(type(l))

[1, 5, 6, 9]
<class 'list'>


In [22]:
# empty tuple

t = ()
print(type(t))

<class 'tuple'>


In [25]:
# tuple

t = (2,4,6)
print(t)
print(type(t))

(2, 4, 6)
<class 'tuple'>


In [226]:
# Empty dict 

d = {}

print(type(d))

<class 'dict'>


In [33]:
# Non empty dictionary

d1 = {1:"python", "two":"machine learning"}
print(d1)
print(type(d1))

{1: 'python', 'two': 'machine learning'}
<class 'dict'>


In [12]:
# set 

set_1 = {}
print(type(set_1))

<class 'dict'>


In [15]:
set_1 = set({})
print(set_1)
print(type(set_1))

set()
<class 'set'>


In [17]:
set_1 = {0,}
print(set_1)
print(type(set_1))

{0}
<class 'set'>


In [ ]:
a = 10
b = 50

c = a + b

 #### operator:-
 
   - Arithmatic operator:-
       - addition (+)
       - subtraction (-)
       - Multiplication (*)
       - Division (/)
       - modulus (%)
       - exponetial (**)
       - floor Division (//)

In [59]:
# addition oprator 

# when data types is numerical 
    
a = int('10')
b = int('50')

print(a+b)  # addition
print(a-b)  # subtraction
print(b-a)  # subtracion

print(a*b) # multiplication

print(b/a) # division (quotient)
print(b%a) # modulus(remainder)
print(b//a) # floor division (integer (quotient))
print(type(b//a))

print(2**2)

60
-40
40
500
5.0
0
5
<class 'int'>
4


In [90]:
# when data types is string

a = 'python'
b = 'programming'

a+ "   " +b
#a-b

#b*5

'python   programming'

In [41]:

(a+b)*2


'pythonprogrammingpythonprogramming'

In [83]:
a = 2+5j
b = 4+5j

print(a+b)
print(a-b)
print(a*b)



(6+10j)
(-2+0j)
(-17+30j)


#### Comparision operator:-
       - equal (==)
       - not equal (!=)
       - greater than (>)
       - less than (<)
       - greather than or equal to (>=)
       - less than or equal to (<=)


In [53]:
a = 10
b = 20
c = 70
d = 10

print( a == b )
print( a == d )

print( a != c)

print( a >= c )

print( a <= c)
print(a >= d)


print(a > c)

False
True
True
False
True
True
False


#### Assignment operator
   - =
   - +=
   - -=
   - *=
   - /=
   - %=
   - **= 
   - //=

In [55]:
a = 10
a

10

In [56]:
a += 10       # a = a + 10
a

20

In [57]:
a -= 5     # a = a - 5
a

15

In [58]:
a /= 3
a

5.0

In [60]:
b = 20

b %= 10

b

0

 #### Logic operator:- 
       - and (Logical AND)
       - or (Logical OR)
       - not (Logical NOT)

In [63]:
a = 20
b = 50 
c = 10
d = 2
e = 10

print( a > b and c == e )  # all condition should be true then it will return true, other wise it will return false
print(a > b or  c == e ) # if any one condition is true then it will return true

print( not c == e )

False
True
False


 #### Membership Operator:-
     this are used to test whether a value or variable is found in a sequence ( String, list, tuple, set and dictionary)
     
       operator                    Meaning
    ---------------------------------------------------
       - in                      True if value/variable is found in the sequence
       - not in                  True if value/varivble is not found in the sequence
       
       
   Note: in a dictionary, we can only test for presence of key not a value

In [66]:
l = [1,5,6,8]

print(1 in l)
print(9 in l)
print(9 not in l)

True
False
True


#### Identity Operator:-
    These are used to check if two value( or variable ) are located on the same part of the memory location of two objects
    
       operator                    Meaning
  ----------------------------------------------------------
       - is                       True if both operands are identical
       - is not                   True if both operands are not identical

In [88]:
A = B =  [5,7,0]
C = [6,9,10]

In [89]:
print(id(A))
print(id(B))
print(id(C))

2335991704392
2335991704392
2335991551304


In [87]:
print(A is B)
print(A is not B)

True
False


In [84]:
print(A is C)
print(A is not B)

False
False


##### What is difference between "is" and "==" operator:-

    This is the operator may seem like the same as the equality operator but they are not same . this 'is' check if both are variable where as the '==' sign checks the value if the variable are same.

In [239]:
A = B = [1,2,3]
C = [1,2,3]

print(id(A))
print(id(B))
print(id(C))

1545151471048
1545151471048
1545154529160


In [78]:
print(A == B)
print(A is B)

True
True


In [79]:
print(A == C)
print(A is C)

False
False


#### Bitwise Operator:
    These are used to perform bit operation. All the decimal value will be converted into binaary value And Bitwise operation will works on these bits such as shifting them left to right or converting value from 0 t0 1 etc.
    
       - & ( Bitwise AND)
       - | ( Bitwise OR)
       - ^ ( Bitwise Exclusive OR)
       - ~ (Bitwise Complement)
       - << (shift Left)
       - '>>' (shift Right)
       
       


##### Bitwise AND (&) :-
 It compares each bit of the first operand to the corressponding bit of the second operand. if both bits are 1 , the corresponding bit is set to 1.
 
 Example:- 
     
    a = 10   ------>  00001010
    b = 20   ------>  00010100
  ------------------------------
    a & b             00000000
    

In [81]:
a = 10 
b = 20
print(a&b)

0


#### Bitwise OR:-
it takes two bit patterns of equal length. this result in each position is 0 if both are 0 while otherwise the result is 1

 Example:- 
     
    a = 10   ------>  00001010
    b = 20   ------>  00010100
  ------------------------------
    a | b             00011110

In [82]:
a = 10 
b = 20
print(a|b)

30


#### Bitwise Exclusive OR
    It compare each bit of its first operand to the corresponding bit of second operand, If one bit is 0 and the other bit is 1 , the corresponding result is set to 1 . otherwise the corresponding result bits is set to 0 ( identical is 0)
    
Examples:-  
         
         a = 1010  
         b = 0011
         --------
       a^b = 1001

In [91]:
a = 10
b = 20
print(a^b)

30


In [93]:
c = 129
d = 156
print(c^d)

29


In [95]:
e = 9
f = 65
print(e^f)

72


#### Binary ones Complement :-

   it is a unary and has the effect of flipping bits
   
example:-

    a = 10  --------> decimal nuber
    a = 0b1010 ---------> binary number
    a = 0b0000 1010 -------> binary 8 bits numbers
    
 Convert '0s' to '1s' and '1s' to '0s' is called Binary ones Complement 
 
 e.g:-
 
    0000 1010   ======> 1111 0101 is Binary ones complement
    
Notes:--
    1. if the converted number starts with 1 the sign is (-)minus
    2. if the converted number starts with 0 the sign is (+)minus

In [2]:
a = 10
print(~a)

-11


In [3]:
b = 20
print(~b)

-21


#### Binary two's complement :-

Adding 1 to Binary one complement is called Binary two Complement.

ex:- 
     
    00001011
           1
    --------------
    00001100

#### << Binary left shift:-
  the left operands value is moved left by the number of bits specified by the right operands
  
  eg:-
  
     a = 10  ---->  1010 ---> 100010
     a<<2   -----> 40 ---> 101000   adding 00 bit 
     

In [8]:
a = 10
print(a<<2)
print(a<<3)

print(bin(40))
print(bin(80))

40
80
0b101000
0b1010000


In [5]:
a = 12
print(a<<3)

96


#### >> Binary right shift:
  The left operands value is moved right by the number specily by the right operands
  
  e.g:-
  
       a = 10 ----> 1010
       a>>2  -----> 10    ----> remove two bits right side   ===> 2

In [11]:
a = 10
print(a>>2)
print(bin(2))

2
0b10


In [12]:
a = 10
print(a>>5)

0


#### Order Of Operator/precidence:-
    
    When an Expression Contains more than one operator, the order of evaluation depends on the order of Operator
    
        operaote                                        Discription
  ------------------------------------------------------------------------------------------------------
          ()                                           Parentheses (grouping)
          **                                           Exponentiation
          ~x                                           Bitwise not
          +x, -x                                       Positive, negative
          *, /, %, //                                  Multiplication, division, remainder, floor division
          +, -	                                     Addition, subtraction
          <<, >>	                                   Bitwise shifts
          &	                                        Bitwise AND
          ^	                                        Bitwise XOR
          |	                                        Bitwise OR
          in, not in, is, is not, <, <=,  
          >,  >=,<>, !=, ==	                        Comparisons, membership, identity
          not x	                                    Boolean NOT
          and	                                      Boolean AND
          or	                                       Boolean OR

In [14]:
a = 20
b = 10
c = 15
d = 5
e = 0

e = (a + b) * c / d       
print("Value of (a + b) * c / d is ",  e)

e = ((a + b) * c) / d     
print( "Value of ((a + b) * c) / d is ",  e)

e = (a + b) * (c / d);    
print("Value of (a + b) * (c / d) is ",  e)

e = a + (b * c) / d;      
print("Value of a + (b * c) / d is ",  e)

Value of (a + b) * c / d is  90.0
Value of ((a + b) * c) / d is  90.0
Value of (a + b) * (c / d) is  90.0
Value of a + (b * c) / d is  50.0


### Python Control Structure:-
    In Python Programming control structure are classified into :-
        - Sequential Control Structure
        - Selection Control Structure
        - Iterative Control Structure

#### Sequential Control Structure:-
    it get executes the line of code in sequential order
    
example:-

In [16]:

print("first line")
print("second line")
print("third line")

first line
second line
third line


In [17]:
print("first");print("second");print("third")

first
second
third


#### Selection Control Statement/Structure:-
    It is Popularly known as Python Decision making. It is anticipation of conditions occuring while execution of an program and specify action taken according to the condition. Python Programming Languages provides Provides following types of decision making statement.
    
    - if                Statement ( one way dicision )
    - if - else         statement ( two way decision )
    - if - elif - else  statement ( multiway decision )
    - nested if -else   statement ( inner dicision )
    - negative condition          (using membership operator )
    
    
##### If statement ( one way decision ):-
    
    it executes a set of statement Conditionally based on the value of a logical expression.
    
    syntax:-
    
        if expression/condition:
            statement-1
            statement-2
            
Examples:-
    

In [20]:
# write a code to take input from the user and check the number is positive or not

num = int(input("Enter a valid number :-"))

if num > 0:
    print("positive number :-", num)
    
print("python")

Enter a valid number :-12
positive number :- 12
python


In [21]:
x = 1
if x>0:
    print("positive")
    
x=-1
if x<0:
    print("Negative")

positive
Negative


#### If else Statement:-

    An else statement can be combined with an if Statement, An else statement contains the block of code of execution. if the condition expression in the if statement false the else block of code will execute.
    
    syntax:- 
    
        if expression:
            statement-1
            statement-2
            
        else:
            statement-3
            statement-4
            
Examples:-

In [23]:
# write a code to take input from the user and check if it is positive , negaative 

x = int(input("Enter any Integer:- "))

if x>0:
    print("positive")
    
else:
    print("negative")

Enter any Integer:- -25
negative


In [25]:
# write a code to check even or odd number

number = int(input("Enter a number :- "))

if number%2 != 0:
    print("odd number")
else:
    print("even number")
    

Enter a number :- 13
odd number


In [24]:
number = int(input("Enter a number :- "))

if number%2 == 0:
    print("Even number")
else:
    print("odd number")
    


Enter a number :- 13
odd number


#### Multiway Condition ( if ...elif ... else):-

    the elif statement allows to check multiple expression for True and execute a block of code as soon as one of the condition evaluation to True
    
    Syntax:-
    
        if Expression:
            
            statement-1
            statement-2
            
        elif Expression:
            
            statement-1
            statement-2
            
        elif Expression:
        
            statement-1
            statement-2
            
         else:
         
             statement-1
             statement-2

Examples:-

In [29]:
# write a code to check the input no. is positive , negative or zero

x = int(input("Enter any Number:- "))

if x>0:
    print("positive")
elif x<0:
    print("negative")
else:
    print("Zero")

Enter any Number:- 0
Zero


In [30]:
x = int(input("Enter any Number:- "))

if x>0:
    print("positive")
if x<0:
    print("negative")
if x == 0:
    print("Zero")

Enter any Number:- 10
positive


In [52]:
# write a code to take a marks from the user and return the grade according to marks and
# above 90   ---- A+ Grade
# 80 ---90   ---- A grade
# 70 -- 80   ---- B grade
# 60 -- 70   ---- B+ grade
# 50 -- 60   ---- C Grade
# 40 -- 50  --- Just pass
# below 40  ------ Fail

marks = int(input("Enter valid marks:- "))

if marks >= 90 and marks <= 100:
    print("Grade A+")
    
elif marks >= 80 and marks <= 90:
    print("Grade A")

elif marks >= 70 and marks <= 80:
    print("Grade B+")
    
elif marks >= 60 and marks <= 70:
    print("Grade B")
    
elif marks >= 50 and marks <= 60:
    print("Grade C")
    
elif marks >= 40 and marks <= 50:
    print("Just pass ")

else:
    print("fail / invalid number")




Enter valid marks:- 12
fail / invalid number


#### Nested Condition or Inner Condition:-

    In General , Nested If ... else statement is used when we want to check more than one condition. Condition are executed from top to bottom and check each condition whether 
    
    
    Syntax:-
    
        if expression:                                               if expression:
        
            statement-1                                                    statement-1
            statement-2                                                    statement-2
            
            if expression:                                           else:
                statement-1                                                statement-1
                statement-2                   or                           statement-2  
                
            else:                                                          if expression:
                statement-1                                                     statement-1
                statement-2                                                     statement-2
                
        else:                                                              else:
            
            statement-1                                                        statement-1
            statement-1                                                        statement-2
            
Examples:-

In [35]:
# write a code to check leap year or not 

year = int(input("Enter valid 4 digit year:- "))



Enter valid 4 digit year:- 2


In [37]:
marks = int(input("inter a valid marks :- "))

if marks >= 90 and marks <= 100:
    print("Grade A+")
    
else:
    if marks >= 80 and marks <= 90:
        print("Grade A")
        
    else:
        if marks >= 70 and marks <= 80:
            print("Grade B+")
        else:
            if marks >= 60 and marks <= 70:
                print("Grade B")
                
            else:
                if marks >= 50 and marks <= 60:
                    print("Grade C")
                else:
                    if marks >= 40 and marks <= 50:
                        print("pass")
                        
                    else:
                        print("Fail")
                    
                
    

inter a valid marks :- 42
pass


#### Python Negative Condition:-
   if the condition is true, the not operator is used to reverse th logic state then logical not operator will make it false
    
    syntax:-
    
        if not expression:
            statement

In [38]:
num = int(input("Enter a number:- "))

if not x>0:
    print("negative number")
else:
    print("positive number")

Enter a number:- 4
positive number


### what is Loop?
    - A loop is a sequence of instructions that is Continuously repeated until a certain condition is reached
   
### Why Loop ?
    - In the loop Structure, the loop ask a question , if the answer requires an action , it is executed , the same question is asking again and again until no further action required, Each time question is asked is called iteration 
   
### Iterative Control Structure:-
    - In General Statements are executed sequentially , the first statement in a function is executed first , followed by the second and so on 
    


Types of loop in python:-
   - for   (finite loop)
   - while (infinite loop)
   - nested for loop
   
   

### For Loop:-
    - It is used to iterate over the items of any sequence inclucing the python list , string , tuple etc.
    
    - syntax:-
        
         for "variable name" in sequence:    # sequence - string, list, tuple, set, frozenset, dictionary
                  statement-1
                  statement-2

Examples :- 

In [39]:
l1 = [1,6,7,4]

for i in l1:
    print(i)

1
6
7
4


In [42]:
str1 = "python"

for i in str1:
    print(i,end=" ")

p y t h o n 

#### Range Function:-
    - It generate lists containing arithmatical process, it returns a list of consecutive integer . the function has three parameters where last two parameters are optional , it is widely used in for loop
    
##### Three variation of range() function:-

   - range(stop)
   - range(start, stop)
   - range(start, stop, step)
   
##### Syntax:-
   - range(a)
   - range(a, b)
   - range(a, b, c)
   
##### range(a):
    - it generates a sequence of numbers from 0 to a , excluding a and incrementing by 1 
    
##### range(a,b):
    - it generates a sequence of numbers from a to b , excluding b and incrementing by 1 
    
##### range(a,b,c):
    - it generates a sequence of numbers from a to b , excluding b and incrementing by c 

Example:-

In [43]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [44]:
for i in range(4,10):
    print(i)

4
5
6
7
8
9


In [46]:
for i in range(1,10,2):
    print(i)

1
3
5
7
9


In [51]:
# printing reverse number

for i in range(10,0,-1):
    print(i)

10
9
8
7
6
5
4
3
2
1


### For Loop with else:-
    A for loop have an optional else block as well. The else part is executed if items in the sequence used in for loop exhausts. Break statement can be used to stop the for loop. in such case  the else part is ignored.
    
    - Syntax:-
         
         for <variable> in <sequence>:
             < statement >
             < statement >
         else:
             < statement >
             < statement >

Example:-

In [4]:
for i in range(5):
    print(i)
else:
    print("else part")
    print(i)

0
1
2
3
4
else part
4


In [5]:
for i in range(5):
    print(i)
    if i == 2:
        break
else:
    print("else part")
    print(i)

0
1
2


### Looping Control Statement :-
    A statement that alerts the execution of a loop from its designation sequence is a loop control Statement 
    
#### Break:- 
    - To Break out from a loop, you can use a keyword "break"   
                             or
    - A break statement inside a loop terminate the loop imidiately
    
    syntax :-
    
        1. break
        
        
        2. for variable in sequence:
                statement-1
                statement-2
                
                if expression:
                    break

Examples:-

In [6]:
for i in range(3,10,2):
    print(i)
    
    if i == 4:
        break

3
5
7
9


In [7]:
for i in range(3,10,1):
    print(i)
    
    if i == 4:
        break

3
4


In [8]:
for i in range(3,10,1):
    if i == 4:
        print(i)
        break

4


In [9]:
for i in range(3,10,1):   
    if i != 4:
        print(i)
        break

3


In [12]:
for i in range(5):
    if i == 5:
        break
    else:
        print(i)  # 0, 1 , 2, 3, 4

0
1
2
3
4


In [14]:
for i in [19, 15, 17, 54]:
    print(i)

19
15
17
54


In [16]:
for i in [19, 15, 17, 54]:
    if i == 15:
        print(i)
        

15


In [18]:
for i in [19, 15, 17, 54]:
     if not i != 15:    # if 19 is not 15 ---- true and not true ----> false
        print(i)        # if 15 is not 15 ---- false and not false ----> true

15


#### continue:-
    - the continue keyword statement is used to tell python interpreter to skip the rest of the statement in the current loop block and to continue to the next iteration if the loop
    
                            or 
                            
    - a continue statement jumps to the next itration of the loop skipping any code in between
    
    syntax:
    
        1. continue
        
        2. for variable_name in sequence:
                statement-1
                statement-2
                
                if expression :
                    continue
                    statement-3
                    statement-4

Examples:-

In [19]:
for i in range(10):
    print(i)
    if i == 5:
        continue

0
1
2
3
4
5
6
7
8
9


In [20]:
for i in range(10):  
    if i == 5:
        print(i)
        continue

5


In [21]:
for i in range(10):  # 0, 1, 2, 3, 4, 5, 6
    if i == 5:
        continue
        print(i)

In [28]:
for i in range(10):
    if i == 5:                    
        continue                  
    else:
        print(i)

IndentationError: unindent does not match any outer indentation level (<tokenize>, line 5)

## While Loop:
    While loop is used to execute number of statements or body till the condition passed in while is True, once the condition is false, the control will come out of the loop
    
    syntax:-
    
        1. while <expression>:
            body
            
            or 
            
            statement

In [66]:
x = 1

while x>0:
    print("positive")
    x = x + 1  # x += 1


positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
positive
p

KeyboardInterrupt: 

In [67]:
x = 1

while x>0:
    print(x)
    
    
    if x == 10:
        break  
    
    x = x + 1  # x += 1


1
2
3
4
5
6
7
8
9
10


In [34]:
x = 1

while x <= 10:
    print(x)
    x = x + 1  # x += 1
    
print(x)


1
2
3
4
5
6
7
8
9
10
11


#### While with else:
    It is same as for with else.
    
    syntax:
          
          while <condition>:
              body
          else:
              body

In [35]:
x = 1

while x <= 10:
    
    print(x)
    x = x + 1  # x += 1
else:  
    print(x)

1
2
3
4
5
6
7
8
9
10
11


### Python Nested Loop:
    python programming language allows to use one loop inside another loop
    
    syntax:
    
        for < first itertive varible > in outer loop:     # row
            [do something ]
            
            for < second itertive varible > in inner loop:  # column
                [do something ]
                
                
#### In nested loop , Find the following facts:-
    - x-axix always contain y-axis
    - y-axix always contain actual data
    - a row contains n number of columns
    - a column contains n number character (data)
    - outer loop represents x-axis (row)
    - inner loop represents y-axis (columns)
    - Mathematically we are representing width and height (x, y)

Examples:-

In [39]:
for i in range(2):
    print(i)
    for j in range(4):
        print(j)

0
0
1
2
3
1
0
1
2
3


In [38]:
for i in range(2):
    for j in range(4):
        print(i, j)

0 0
0 1
0 2
0 3
1 0
1 1
1 2
1 3


In [40]:
for i in range(2):
    for j in range(4):
        if i == 0 and j == 3:
            print(i, j)

0 3


In [42]:
for i in range(2):
    for j in range(4):
        if i == 0:
            print(i, j)
            

0 0
0 1
0 2
0 3


In [43]:
for i in range(2):
    for j in range(4):
        if i == 1:
            print(i, j)
            

1 0
1 1
1 2
1 3


In [44]:
for i in range(2):  #---> rows
    for j in range(4): # ----> columns
        if i == 1 and j > 0:
            print(i, j)
            

1 1
1 2
1 3


# Assignments :--

In [ ]:
# write a code to binary sum of 9 and 10  ----> 19    # assignment

## Patterns

In [60]:
for i in range(1,5+1):
    for j in range(1,i+1):
        print(j, end=" ")
    print()
        
        

1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


In [15]:
for i in range(1,5+1):
    for j in range(1,i+1):
        print(i, end=" ")
    print()

1 
2 2 
3 3 3 
4 4 4 4 
5 5 5 5 5 


In [1]:
n = int(input("Enter a number :- "))

for i in range(1, n+1):
    print("  "*(n-i), end="")
    for j in range(1,i+1):
        print(j, end=" ")
    print()

Enter a number :- 6
          1 
        1 2 
      1 2 3 
    1 2 3 4 
  1 2 3 4 5 
1 2 3 4 5 6 


In [2]:
n = int(input("Enter a number :- "))

for i in range(1, n+1):
    print("  "*(n-i), end="")
    for j in range(1,i+1):
        print(i, end=" ")
    print()

Enter a number :- 7
            1 
          2 2 
        3 3 3 
      4 4 4 4 
    5 5 5 5 5 
  6 6 6 6 6 6 
7 7 7 7 7 7 7 


In [19]:
n = int(input("Enter a number :- "))

for i in range(n,0,-1):
    for j in range(1,i+1):
        print(j,end=" ")
        
    print()

Enter a number :- 7
1 2 3 4 5 6 7 
1 2 3 4 5 6 
1 2 3 4 5 
1 2 3 4 
1 2 3 
1 2 
1 


In [5]:
n = int(input("Enter a number :- "))

for i in range(n,0,-1):
    for j in range(i,0,-1):
        print(j,end=" ")  
    print()

Enter a number :- 5
5 4 3 2 1 
4 3 2 1 
3 2 1 
2 1 
1 


In [6]:
n = int(input("Enter a number :- "))

for i in range(n,0,-1):
    print("  "*(n-i), end=" ")
    for j in range(1,i+1):
        print(j,end=" ")
    print()
    

Enter a number :- 5
 1 2 3 4 5 
   1 2 3 4 
     1 2 3 
       1 2 
         1 


In [5]:
for r in range(6,0,-1):
    print("  "*(6-r), end=" ")
    for c in range(r,0,-1):
        print(c, end=" ")
        
    print()

 6 5 4 3 2 1 
   5 4 3 2 1 
     4 3 2 1 
       3 2 1 
         2 1 
           1 


In [7]:
n = int(input("Enter a number :- "))

for i in range(1,n+1):
    for j in range(i):
        print(n-j, end=" ")
    print()

Enter a number :- 5
5 
5 4 
5 4 3 
5 4 3 2 
5 4 3 2 1 


In [18]:
n = int(input("Enter a value:-"))

for i in range(n,0,-1):
    print("  "*(n-i), end='')
    for j in range(i,0,-1):
        print(n-(i-j), end=' ')
        #print((i-j), end=" ")
    print()
        

Enter a value:-5
5 4 3 2 1 
  5 4 3 2 
    5 4 3 
      5 4 
        5 


In [23]:
n = int(input("Enter a number:- "))

for i in range(1,n+1):
    print("  "*(n-i), end='')
    
    for j in range(1,i+1):
        print(n-(i-j), end=' ')
        
    for k in range(i-1,0,-1):
        print(k+(n-i), end=' ')
        
    print()

Enter a number:- 5
        5 
      4 5 4 
    3 4 5 4 3 
  2 3 4 5 4 3 2 
1 2 3 4 5 4 3 2 1 


In [26]:
n = int(input("Enter a number:- "))

for i in range(1,n+1):
    print("  "*(n-i), end='')
    
    for j in range(1,i+1):
        print(n-(i-j), end=' ')
        
    for k in range(i-1,0,-1):
        print(k+(n-i), end=' ')
        
    print()
    
for i in range(n-1,0,-1):
    print("  "*(n-i), end='')
    
    for j in range(1,i+1):
        print(n-(i-j), end=' ')
        
    for k in range(i-1,0,-1):
        print(k+(n-i), end=' ')
        
    print()

Enter a number:- 6
          6 
        5 6 5 
      4 5 6 5 4 
    3 4 5 6 5 4 3 
  2 3 4 5 6 5 4 3 2 
1 2 3 4 5 6 5 4 3 2 1 
  2 3 4 5 6 5 4 3 2 
    3 4 5 6 5 4 3 
      4 5 6 5 4 
        5 6 5 
          6 


In [30]:
n = int(input("Inter a number:- "))

for i in range(1, n ):
    for j in range(1,i+1):
        print(chr(j+96), end=' ')
    print()

Inter a number:- 5
a 
a b 
a b c 
a b c d 


In [11]:
print(ord("a"))
print(chr(97))

97
a


In [12]:
n = int(input("Enter any number:-"))

for i in range(n):
    for j in range(1,i+1):
        print(chr(j+96), end=" ")
    print()

Enter any number:-6

a 
a b 
a b c 
a b c d 
a b c d e 


In [ ]:
write a code to print the following pattern.


a b c d
a b c
a b
a

In [33]:
n = int(input("Enter any number:-"))

for i in range(n):
    for j in range(1,i+1):
        print('*', end=" ")
    print()

Enter any number:-5

* 
* * 
* * * 
* * * * 


In [4]:
number = 15

binary_number = int("{0:08b}".format(number))

flipped_binary_number = ~ binary_number

flipped_binary_number = flipped_binary_number + 1

str_twos_complement = str(flipped_binary_number)

twos_complement = int(str_twos_complement, 2)

print(twos_complement)


-15


In [35]:
num = 20
print(bin(20)[2:])
binary_number = int("{0:08b}".format(num))
print(binary_number)

flipped_binary_number = ~ binary_number
print(flipped_binary_number)

flipped_binary_number = flipped_binary_number + 1
print(flipped_binary_number)

str_twos_complement = str(flipped_binary_number)
print(str_twos_complement)

twos_complement = int(str_twos_complement, 2)

print(twos_complement)

10100
10100
-10101
-10100
-10100
-20


In [41]:
if "uei".isalpha():
    print("bangalore")

bangalore


In [ ]:
0 ---> False ----- None ==== > not printing any things
